src/groupeddataframe/grouping.jl

```diff
@@ -412,28 +412,60 @@ true
 """
 Base.keys(gd::GroupedDataFrame) = GroupKeys(gd)
 
-# Index with GroupKey
-function Base.getindex(gd::GroupedDataFrame, key::GroupKey)
-    gd === parent(key) && return gd[getfield(key, :idx)]
+
+# The allowed key types for dictionary-like indexing
+const GroupKeyTypes = Union{GroupKey, Tuple, NamedTuple}
+# All allowed scalar index types
+const GroupIndexTypes = Union{Integer, GroupKeyTypes}
+
+
+# Find the integer index of a group given any supported scalar index type
+_findgroup(gd::GroupedDataFrame, i::Integer) = i
+
+function _findgroup(gd::GroupedDataFrame, key::GroupKey)
+    gd === parent(key) && return getfield(key, :idx)
     throw(ErrorException("Cannot use a GroupKey to index a GroupedDataFrame other than the one it was derived from."))
 end
 
-# Index with tuple
-function Base.getindex(gd::GroupedDataFrame, key::Tuple)
+function _findgroup(gd::GroupedDataFrame, key::Tuple)
     for i in 1:length(gd)
-        isequal(Tuple(_groupvalues(gd, i)), key) && return gd[i]
+        isequal(Tuple(_groupvalues(gd, i)), key) && return i
     end
     throw(KeyError(key))
 end
 
-# Index with named tuple
-function Base.getindex(gd::GroupedDataFrame, key::NamedTuple{N}) where {N}
+function _findgroup(gd::GroupedDataFrame, key::NamedTuple{N}) where {N}
     if length(key) != length(gd.cols) || any(n != _names(gd)[c] for (n, c) in zip(N, gd.cols))
         throw(KeyError(key))
     end
-    return gd[Tuple(key)]
+    return _findgroup(gd, values(key))
+end
+
+# Dictionary-like indexing
+Base.getindex(gd::GroupedDataFrame, key::GroupKeyTypes) = gd[_findgroup(gd, key)]
+
+# Array indexing with dictionary-like keys
+Base.getindex(gd::GroupedDataFrame, idxs::AbstractVector{T}) where {T<:GroupKeyTypes} = gd[[_findgroup(gd, k) for k in idxs]]
+
+# InvertedIndex wrapping scalar index
+function Base.getindex(gd::GroupedDataFrame, idx::InvertedIndex{T}) where {T<:GroupIndexTypes}
+    selected = trues(length(gd))
+    selected[_findgroup(gd, idx.skip)] = false
```

Comment:
I would think this would be faster - the existing `getindex(gd, ::AbstractArray)` works directly with either ints or bools because the argument is just used as an index for `.starts` and `.ends`.  This way is a bit array allocation and an `O(1)` write, `setdiff` allocates an integer array but also has to do an `O(n)` scan through all the indices.

Reply:
My point is that in the end `getindex` on vectors will actually do an `O(n)` pass anyway to count the number of `true` entries (and probably twice, for `starts` and `ends`). Since you know that value, it should be better to directly allocate an array of integer indices. Maybe try something like `n = _findgroup(gd, idx.skip); vcat(1:n, n+2:length(gd))`?

Reply:
Ah, but that will only work for single indices. I guess it's OK to keep a logical index then. (Ideally we would have an optimized `setdiff` method for ranges, but...)

Reply:
I see what you mean. Yes I think that's probably the best way of doing it.